# CaSR to Operational Forcing Fields

This notebook converts an integrated `CaSR` forcing field into an operational forcing field suitable for `SalishSeaCast`.

## CaSR information

CaSR file names:

    Integrated_RDPS_P_FB_SFC_2008_2024.nc   Integrated_RDPS_P_PR0_SFC_2008_2024.nc
    Integrated_RDPS_P_FI_SFC_2008_2024.nc   Integrated_RDPS_P_TT_1.5m_2008_2024.nc
    Integrated_RDPS_P_HU_1.5m_2008_2024.nc  Integrated_RDPS_P_UUC_10m_2008_2024.nc
    Integrated_RDPS_P_PN_SFC_2008_2024.nc   Integrated_RDPS_P_VVC_10m_2008_2024.nc

Base path: `/ocean/jqiu/Atmospheric_RDPS/2008_2024_Integrated/`

In [1]:
# paths

base_path_casr = '/ocean/jqiu/Atmospheric_RDPS/2008_2024_Integrated/'

fname_FB  = 'Integrated_RDPS_P_FB_SFC_2008_2024.nc'
fname_PRO = 'Integrated_RDPS_P_PR0_SFC_2008_2024.nc'
fname_FI  = 'Integrated_RDPS_P_FI_SFC_2008_2024.nc'
fname_TT  = 'Integrated_RDPS_P_TT_1.5m_2008_2024.nc'
fname_HU  = 'Integrated_RDPS_P_HU_1.5m_2008_2024.nc'
fname_UUC = 'Integrated_RDPS_P_UUC_10m_2008_2024.nc'
fname_PN  = 'Integrated_RDPS_P_PN_SFC_2008_2024.nc'
fname_VVC = 'Integrated_RDPS_P_VVC_10m_2008_2024.nc'

path_FB  = base_path_casr + fname_FB
path_PRO = base_path_casr + fname_PRO
path_FI  = base_path_casr + fname_FI
path_TT  = base_path_casr + fname_TT
path_HU  = base_path_casr + fname_HU
path_UUC = base_path_casr + fname_UUC
path_PN  = base_path_casr + fname_PN
path_VVC = base_path_casr + fname_VVC

In [2]:
import sys
sys.path.append("/home/jqiu/analysis-junqi/Tools-Junqi")

import junqi_nctool as jqnc

### SFC variables

In [3]:
jqnc.nc_disp(path_FB,detailed=True)

### NetCDF Summary

*Path: : `/ocean/jqiu/Atmospheric_RDPS/2008_2024_Integrated/Integrated_RDPS_P_FB_SFC_2008_2024.nc`*

#### Dimensions

Dimension,Size
time,149040
rlat,105
rlon,70


#### Variables

Name,Shape,Dimensions,Long Name,Units
rotated_pole,(),scalar,,
CaSR_v3.2_P_FB_SFC,"(149040, 105, 70)","time, rlat, rlon",Forecast: Downward solar flux at surface,W m**-2
time,"(149040,)",time,Validity time,
lon,"(105, 70)","rlat, rlon",longitude,degrees_east
lat,"(105, 70)","rlat, rlon",latitude,degrees_north
rlon,"(70,)",rlon,longitude in rotated pole grid,degrees
rlat,"(105,)",rlat,latitude in rotated pole grid,degrees


#### Global Attributes

Attribute,Value
Conventions,CF-1.6
source,https://hpfx.collab.science.gc.ca/~scar700/rcas-casr/
institution,Environment and Climate Change Canada / Environnement et Changement climatique Canada
title,Canadian Surface Reanalysis (CaSR) / Réanalyse Canadienne de Surface (RCaS)
contact,rcas-casr@ec.gc.ca
licence,These data are produced and provided by the Environment and Climate Change Canada. License agreement can be found in https://open.canada.ca/en/open-government-licence-canada or email rcas-casr@ec.gc.ca / Ces données sont produites et fournies par Environnement et Changement climatique Canada. Le contrat de license peut être trouvé dans https://ouvert.canada.ca/fr/licence-du-gouvernement-ouvert-canada ou courriel rcas-casr@ec.gc.ca


In [4]:
jqnc.nc_disp(path_PRO,detailed=True)

### NetCDF Summary

*Path: : `/ocean/jqiu/Atmospheric_RDPS/2008_2024_Integrated/Integrated_RDPS_P_PR0_SFC_2008_2024.nc`*

#### Dimensions

Dimension,Size
time,149040
rlat,105
rlon,70


#### Variables

Name,Shape,Dimensions,Long Name,Units
rotated_pole,(),scalar,,
CaSR_v3.2_P_PR0_SFC,"(149040, 105, 70)","time, rlat, rlon",Forecast: Quantity of precipitation (model) at surface,m
time,"(149040,)",time,Validity time,
lon,"(105, 70)","rlat, rlon",longitude,degrees_east
lat,"(105, 70)","rlat, rlon",latitude,degrees_north
rlon,"(70,)",rlon,longitude in rotated pole grid,degrees
rlat,"(105,)",rlat,latitude in rotated pole grid,degrees


#### Global Attributes

Attribute,Value
Conventions,CF-1.6
source,https://hpfx.collab.science.gc.ca/~scar700/rcas-casr/
institution,Environment and Climate Change Canada / Environnement et Changement climatique Canada
title,Canadian Surface Reanalysis (CaSR) / Réanalyse Canadienne de Surface (RCaS)
contact,rcas-casr@ec.gc.ca
licence,These data are produced and provided by the Environment and Climate Change Canada. License agreement can be found in https://open.canada.ca/en/open-government-licence-canada or email rcas-casr@ec.gc.ca / Ces données sont produites et fournies par Environnement et Changement climatique Canada. Le contrat de license peut être trouvé dans https://ouvert.canada.ca/fr/licence-du-gouvernement-ouvert-canada ou courriel rcas-casr@ec.gc.ca


In [5]:
jqnc.nc_disp(path_FI,detailed=True)

### NetCDF Summary

*Path: : `/ocean/jqiu/Atmospheric_RDPS/2008_2024_Integrated/Integrated_RDPS_P_FI_SFC_2008_2024.nc`*

#### Dimensions

Dimension,Size
time,149040
rlat,105
rlon,70


#### Variables

Name,Shape,Dimensions,Long Name,Units
rotated_pole,(),scalar,,
CaSR_v3.2_P_FI_SFC,"(149040, 105, 70)","time, rlat, rlon",Forecast: Surface incoming infrared flux at surface,W m**-2
time,"(149040,)",time,Validity time,
lon,"(105, 70)","rlat, rlon",longitude,degrees_east
lat,"(105, 70)","rlat, rlon",latitude,degrees_north
rlon,"(70,)",rlon,longitude in rotated pole grid,degrees
rlat,"(105,)",rlat,latitude in rotated pole grid,degrees


#### Global Attributes

Attribute,Value
Conventions,CF-1.6
source,https://hpfx.collab.science.gc.ca/~scar700/rcas-casr/
institution,Environment and Climate Change Canada / Environnement et Changement climatique Canada
title,Canadian Surface Reanalysis (CaSR) / Réanalyse Canadienne de Surface (RCaS)
contact,rcas-casr@ec.gc.ca
licence,These data are produced and provided by the Environment and Climate Change Canada. License agreement can be found in https://open.canada.ca/en/open-government-licence-canada or email rcas-casr@ec.gc.ca / Ces données sont produites et fournies par Environnement et Changement climatique Canada. Le contrat de license peut être trouvé dans https://ouvert.canada.ca/fr/licence-du-gouvernement-ouvert-canada ou courriel rcas-casr@ec.gc.ca


In [6]:
jqnc.nc_disp(path_PN,detailed=True)

### NetCDF Summary

*Path: : `/ocean/jqiu/Atmospheric_RDPS/2008_2024_Integrated/Integrated_RDPS_P_PN_SFC_2008_2024.nc`*

#### Dimensions

Dimension,Size
time,149040
rlat,105
rlon,70


#### Variables

Name,Shape,Dimensions,Long Name,Units
rotated_pole,(),scalar,,
CaSR_v3.2_P_PN_SFC,"(149040, 105, 70)","time, rlat, rlon",Forecast: Sea level pressure at surface,mb
time,"(149040,)",time,Validity time,
lon,"(105, 70)","rlat, rlon",longitude,degrees_east
lat,"(105, 70)","rlat, rlon",latitude,degrees_north
rlon,"(70,)",rlon,longitude in rotated pole grid,degrees
rlat,"(105,)",rlat,latitude in rotated pole grid,degrees


#### Global Attributes

Attribute,Value
Conventions,CF-1.6
source,https://hpfx.collab.science.gc.ca/~scar700/rcas-casr/
institution,Environment and Climate Change Canada / Environnement et Changement climatique Canada
title,Canadian Surface Reanalysis (CaSR) / Réanalyse Canadienne de Surface (RCaS)
contact,rcas-casr@ec.gc.ca
licence,These data are produced and provided by the Environment and Climate Change Canada. License agreement can be found in https://open.canada.ca/en/open-government-licence-canada or email rcas-casr@ec.gc.ca / Ces données sont produites et fournies par Environnement et Changement climatique Canada. Le contrat de license peut être trouvé dans https://ouvert.canada.ca/fr/licence-du-gouvernement-ouvert-canada ou courriel rcas-casr@ec.gc.ca


### 1.5 m variables

In [7]:
jqnc.nc_disp(path_TT,detailed=True)

### NetCDF Summary

*Path: : `/ocean/jqiu/Atmospheric_RDPS/2008_2024_Integrated/Integrated_RDPS_P_TT_1.5m_2008_2024.nc`*

#### Dimensions

Dimension,Size
time,149040
rlat,105
rlon,70


#### Variables

Name,Shape,Dimensions,Long Name,Units
rotated_pole,(),scalar,,
CaSR_v3.2_P_TT_1.5m,"(149040, 105, 70)","time, rlat, rlon",Forecast: Air temperature at 1.5m,deg_C
time,"(149040,)",time,Validity time,
lon,"(105, 70)","rlat, rlon",longitude,degrees_east
lat,"(105, 70)","rlat, rlon",latitude,degrees_north
rlon,"(70,)",rlon,longitude in rotated pole grid,degrees
rlat,"(105,)",rlat,latitude in rotated pole grid,degrees


#### Global Attributes

Attribute,Value
Conventions,CF-1.6
source,https://hpfx.collab.science.gc.ca/~scar700/rcas-casr/
institution,Environment and Climate Change Canada / Environnement et Changement climatique Canada
title,Canadian Surface Reanalysis (CaSR) / Réanalyse Canadienne de Surface (RCaS)
contact,rcas-casr@ec.gc.ca
licence,These data are produced and provided by the Environment and Climate Change Canada. License agreement can be found in https://open.canada.ca/en/open-government-licence-canada or email rcas-casr@ec.gc.ca / Ces données sont produites et fournies par Environnement et Changement climatique Canada. Le contrat de license peut être trouvé dans https://ouvert.canada.ca/fr/licence-du-gouvernement-ouvert-canada ou courriel rcas-casr@ec.gc.ca


In [8]:
jqnc.nc_disp(path_HU,detailed=True)

### NetCDF Summary

*Path: : `/ocean/jqiu/Atmospheric_RDPS/2008_2024_Integrated/Integrated_RDPS_P_HU_1.5m_2008_2024.nc`*

#### Dimensions

Dimension,Size
time,149040
rlat,105
rlon,70


#### Variables

Name,Shape,Dimensions,Long Name,Units
rotated_pole,(),scalar,,
CaSR_v3.2_P_HU_1.5m,"(149040, 105, 70)","time, rlat, rlon",Forecast: Specific humidity at 1.5m,kg kg**-1
time,"(149040,)",time,Validity time,
lon,"(105, 70)","rlat, rlon",longitude,degrees_east
lat,"(105, 70)","rlat, rlon",latitude,degrees_north
rlon,"(70,)",rlon,longitude in rotated pole grid,degrees
rlat,"(105,)",rlat,latitude in rotated pole grid,degrees


#### Global Attributes

Attribute,Value
Conventions,CF-1.6
source,https://hpfx.collab.science.gc.ca/~scar700/rcas-casr/
institution,Environment and Climate Change Canada / Environnement et Changement climatique Canada
title,Canadian Surface Reanalysis (CaSR) / Réanalyse Canadienne de Surface (RCaS)
contact,rcas-casr@ec.gc.ca
licence,These data are produced and provided by the Environment and Climate Change Canada. License agreement can be found in https://open.canada.ca/en/open-government-licence-canada or email rcas-casr@ec.gc.ca / Ces données sont produites et fournies par Environnement et Changement climatique Canada. Le contrat de license peut être trouvé dans https://ouvert.canada.ca/fr/licence-du-gouvernement-ouvert-canada ou courriel rcas-casr@ec.gc.ca


### 10 m variables

In [9]:
jqnc.nc_disp(path_UUC,detailed=True)

### NetCDF Summary

*Path: : `/ocean/jqiu/Atmospheric_RDPS/2008_2024_Integrated/Integrated_RDPS_P_UUC_10m_2008_2024.nc`*

#### Dimensions

Dimension,Size
time,149040
rlat,105
rlon,70


#### Variables

Name,Shape,Dimensions,Long Name,Units
rotated_pole,(),scalar,,
CaSR_v3.2_P_UUC_10m,"(149040, 105, 70)","time, rlat, rlon",Forecast: Corrected U-component of the wind (along West-East direction) at 10m,kts
time,"(149040,)",time,Validity time,
lon,"(105, 70)","rlat, rlon",longitude,degrees_east
lat,"(105, 70)","rlat, rlon",latitude,degrees_north
rlon,"(70,)",rlon,longitude in rotated pole grid,degrees
rlat,"(105,)",rlat,latitude in rotated pole grid,degrees


#### Global Attributes

Attribute,Value
Conventions,CF-1.6
source,https://hpfx.collab.science.gc.ca/~scar700/rcas-casr/
institution,Environment and Climate Change Canada / Environnement et Changement climatique Canada
title,Canadian Surface Reanalysis (CaSR) / Réanalyse Canadienne de Surface (RCaS)
contact,rcas-casr@ec.gc.ca
licence,These data are produced and provided by the Environment and Climate Change Canada. License agreement can be found in https://open.canada.ca/en/open-government-licence-canada or email rcas-casr@ec.gc.ca / Ces données sont produites et fournies par Environnement et Changement climatique Canada. Le contrat de license peut être trouvé dans https://ouvert.canada.ca/fr/licence-du-gouvernement-ouvert-canada ou courriel rcas-casr@ec.gc.ca


In [10]:
jqnc.nc_disp(path_VVC,detailed=True)

### NetCDF Summary

*Path: : `/ocean/jqiu/Atmospheric_RDPS/2008_2024_Integrated/Integrated_RDPS_P_VVC_10m_2008_2024.nc`*

#### Dimensions

Dimension,Size
time,149040
rlat,105
rlon,70


#### Variables

Name,Shape,Dimensions,Long Name,Units
rotated_pole,(),scalar,,
CaSR_v3.2_P_VVC_10m,"(149040, 105, 70)","time, rlat, rlon",Forecast: Corrected V-component of the wind (along South-North direction) at 10m,kts
time,"(149040,)",time,Validity time,
lon,"(105, 70)","rlat, rlon",longitude,degrees_east
lat,"(105, 70)","rlat, rlon",latitude,degrees_north
rlon,"(70,)",rlon,longitude in rotated pole grid,degrees
rlat,"(105,)",rlat,latitude in rotated pole grid,degrees


#### Global Attributes

Attribute,Value
Conventions,CF-1.6
source,https://hpfx.collab.science.gc.ca/~scar700/rcas-casr/
institution,Environment and Climate Change Canada / Environnement et Changement climatique Canada
title,Canadian Surface Reanalysis (CaSR) / Réanalyse Canadienne de Surface (RCaS)
contact,rcas-casr@ec.gc.ca
licence,These data are produced and provided by the Environment and Climate Change Canada. License agreement can be found in https://open.canada.ca/en/open-government-licence-canada or email rcas-casr@ec.gc.ca / Ces données sont produites et fournies par Environnement et Changement climatique Canada. Le contrat de license peut être trouvé dans https://ouvert.canada.ca/fr/licence-du-gouvernement-ouvert-canada ou courriel rcas-casr@ec.gc.ca


### Lon and Lat 

In [ ]:
import xarray as xr
import os

# 拼接完整的文件路径
base_path_casr = '/ocean/jqiu/Atmospheric_RDPS/2008_2024_Integrated/'
fname_FB  = 'Integrated_RDPS_P_FB_SFC_2008_2024.nc'
file_path = os.path.join(base_path_casr, fname_FB)

# 读取 NetCDF 文件
try:
    # 默认引擎通常是 netcdf4或h5netcdf
    ds = xr.open_dataset(file_path)
    
    # 获取 lon 和 lat，并计算它们的最小值和最大值
    # .item() 用于将单个元素的 xarray DataArray 转换为 Python 原生的 float 类型
    lon_min = ds['lon'].min().item()
    lon_max = ds['lon'].max().item()
    
    lat_min = ds['lat'].min().item()
    lat_max = ds['lat'].max().item()
    
    # 打印结果
    print(f"文件: {fname_FB}")
    print("-" * 30)
    print(f"经度 (lon) 范围: {lon_min:.4f}  到  {lon_max:.4f}")
    print(f"纬度 (lat) 范围: {lat_min:.4f}  到  {lat_max:.4f}")
    
    # 关闭数据集释放内存
    ds.close()

except Exception as e:
    print(f"读取文件时出错: {e}")

文件: Integrated_RDPS_P_FB_SFC_2008_2024.nc
------------------------------
经度 (lon) 范围: 229.0648  到  243.7545
纬度 (lat) 范围: 43.3772  到  54.4950


## HRDPS information

In [12]:
path_hrdps='/results/forcing/atmospheric/GEM2.5/operational/ops_y2018m01d01.nc'

jqnc.nc_disp(path_hrdps,detailed=True)

### NetCDF Summary

*Path: : `/results/forcing/atmospheric/GEM2.5/operational/ops_y2018m01d01.nc`*

#### Dimensions

Dimension,Size
time_counter,24
y,266
x,256


#### Variables

Name,Shape,Dimensions,Long Name,Units
atmpres,"(24, 266, 256)","time_counter, y, x",Pressure Reduced to MSL,Pa
nav_lat,"(266, 256)","y, x",latitude,degrees_north
nav_lon,"(266, 256)","y, x",longitude,degrees_east
percentcloud,"(24, 266, 256)","time_counter, y, x",Total Cloud Cover,percent
precip,"(24, 266, 256)","time_counter, y, x",Total Precipitation,kg/m^2
qair,"(24, 266, 256)","time_counter, y, x",Specific Humidity,kg/kg
solar,"(24, 266, 256)","time_counter, y, x",Downward Short-Wave Radiation Flux,W/m^2
tair,"(24, 266, 256)","time_counter, y, x",Temperature,K
therm_rad,"(24, 266, 256)","time_counter, y, x",Downward Long-Wave Rad. Flux,W/m^2
u_wind,"(24, 266, 256)","time_counter, y, x",U-Component of Wind,m/s


#### Global Attributes

Attribute,Value
Conventions,CF-1.0
History,Mon Jan 1 10:36:51 2018: ncks -4 -L4 -O /results/forcing/atmospheric/GEM2.5/operational/ops_y2018m01d01.nc /results/forcing/atmospheric/GEM2.5/operational/ops_y2018m01d01.nc created by wgrib2
GRIB2_grid_template,20
NCO,4.4.2


In [5]:
import xarray as xr
import os

# 文件路径

file_path = '/results/forcing/atmospheric/GEM2.5/operational/ops_y2018m01d01.nc'

# 读取 NetCDF 文件
try:
    # 默认引擎通常是 netcdf4或h5netcdf
    ds = xr.open_dataset(file_path)
    
    # 获取 lon 和 lat，并计算它们的最小值和最大值
    # .item() 用于将单个元素的 xarray DataArray 转换为 Python 原生的 float 类型
    lon_min = ds['nav_lon'].min().item()
    lon_max = ds['nav_lon'].max().item()
    
    lat_min = ds['nav_lat'].min().item()
    lat_max = ds['nav_lat'].max().item()
    
    # 打印结果
    print(f"文件: {fname_FB}")
    print("-" * 30)
    print(f"经度 (lon) 范围: {lon_min:.4f}  到  {lon_max:.4f}")
    print(f"纬度 (lat) 范围: {lat_min:.4f}  到  {lat_max:.4f}")
    
    # 关闭数据集释放内存
    ds.close()

except Exception as e:
    print(f"读取文件时出错: {e}")

print(ds.time_counter)

文件: Integrated_RDPS_P_FB_SFC_2008_2024.nc
------------------------------
经度 (lon) 范围: 230.8331  到  240.5296
纬度 (lat) 范围: 45.6649  到  52.1793
<xarray.DataArray 'time_counter' (time_counter: 24)> Size: 192B
array(['2018-01-01T00:00:00.000000000', '2018-01-01T01:00:00.000000000',
       '2018-01-01T02:00:00.000000000', '2018-01-01T03:00:00.000000000',
       '2018-01-01T04:00:00.000000000', '2018-01-01T05:00:00.000000000',
       '2018-01-01T06:00:00.000000000', '2018-01-01T07:00:00.000000000',
       '2018-01-01T08:00:00.000000000', '2018-01-01T09:00:00.000000000',
       '2018-01-01T10:00:00.000000000', '2018-01-01T11:00:00.000000000',
       '2018-01-01T12:00:00.000000000', '2018-01-01T13:00:00.000000000',
       '2018-01-01T14:00:00.000000000', '2018-01-01T15:00:00.000000000',
       '2018-01-01T16:00:00.000000000', '2018-01-01T17:00:00.000000000',
       '2018-01-01T18:00:00.000000000', '2018-01-01T19:00:00.000000000',
       '2018-01-01T20:00:00.000000000', '2018-01-01T21:00:00.0000

### Summary

It seems that `Total Cloud Cover` is missing and I believe that it's not provided by `CaSR`, but that variable is not necessary. See https://hpfx.collab.science.gc.ca/~scar700/rcas-casr/download_CaSR_regions_var_period.html



| CaSR variable description | CaSR units | HRDPS variable name | HRDPS long name | HRDPS units | Notes |
|---|---:|---|---|---:|---|
| Forecast: Corrected V-component of the wind along South-North direction at 10 m | kts | v_wind | V-Component of Wind | m/s  | Same wind component, but units differ: CaSR uses knots, HRDPS uses m/s. |
| Forecast: Corrected U-component of the wind along West-East direction at 10 m | kts | u_wind | U-Component of Wind | m/s  | Same wind component, but units differ: CaSR uses knots, HRDPS uses m/s. |
| Forecast: Specific humidity at 1.5 m | kg kg**-1 | qair | Specific Humidity | kg/kg  | Units are effectively equivalent. |
| Forecast: Air temperature at 1.5 m | deg_C | tair | Temperature | K  | Same temperature variable, but units differ: CaSR uses Celsius, HRDPS uses Kelvin. |
| Forecast: Sea level pressure at surface | mb | atmpres | Pressure Reduced to MSL | Pa  | Same sea-level pressure concept, but units differ: CaSR uses millibars, HRDPS uses pascals. |
| Forecast: Surface incoming infrared flux at surface | W m**-2 | therm_rad | Downward Long-Wave Rad. Flux | W/m^2  | Incoming infrared flux corresponds to downward long-wave radiation. |
| Forecast: Quantity of precipitation model at surface | m | precip | Total Precipitation | kg/m^2  | Same precipitation-related variable, but units differ. For water, 1 kg/m² is approximately 1 mm water equivalent. |
| Forecast: Downward solar flux at surface | W m**-2 | solar | Downward Short-Wave Radiation Flux | W/m^2  | Downward solar flux corresponds to downward short-wave radiation. |
|Not Provided | | percentcloud | Total Cloud Cover | percent | This variable appears in HRDPS but was not included in the provided CaSR variable list.






## Test Conversion and inspection

Try to generate one day of data, and see if there are any mistakes.

In [1]:
import xarray as xr
import numpy as np
import pandas as pd
import os

# 基础路径与配置
base_path = '/ocean/jqiu/Atmospheric_RDPS/2008_2024_Integrated/'
target_date = '2023-03-01' # 先截取 2023 年 3 月 1 日进行测试

# 目标变量名 : (原文件名, 原变量名)
var_mapping = {
    'solar':     ('Integrated_RDPS_P_FB_SFC_2008_2024.nc',  'CaSR_v3.2_P_FB_SFC'),
    'precip':    ('Integrated_RDPS_P_PR0_SFC_2008_2024.nc', 'CaSR_v3.2_P_PR0_SFC'),
    'therm_rad': ('Integrated_RDPS_P_FI_SFC_2008_2024.nc',  'CaSR_v3.2_P_FI_SFC'),
    'tair':      ('Integrated_RDPS_P_TT_1.5m_2008_2024.nc', 'CaSR_v3.2_P_TT_1.5m'),
    'qair':      ('Integrated_RDPS_P_HU_1.5m_2008_2024.nc', 'CaSR_v3.2_P_HU_1.5m'),
    'u_wind':    ('Integrated_RDPS_P_UUC_10m_2008_2024.nc', 'CaSR_v3.2_P_UUC_10m'),
    'atmpres':   ('Integrated_RDPS_P_PN_SFC_2008_2024.nc',  'CaSR_v3.2_P_PN_SFC'),
    'v_wind':    ('Integrated_RDPS_P_VVC_10m_2008_2024.nc', 'CaSR_v3.2_P_VVC_10m')
}

# 初始化目标 Dataset 与坐标
ds_out = xr.Dataset()

# 用 TT(温度) 文件作为参考来获取网格和时间信息
ref_file = os.path.join(base_path, 'Integrated_RDPS_P_TT_1.5m_2008_2024.nc')
ds_ref = xr.open_dataset(ref_file).sel(time=target_date)

y_size, x_size = ds_ref.rlat.size, ds_ref.rlon.size
time_size = ds_ref.time.size  # 获取实际的时间步数量，替代硬编码的 24

# 建立目标格式要求的坐标系
ds_out.coords['y'] = np.arange(y_size)
ds_out.coords['x'] = np.arange(x_size)
# 【修改点 1】直接提取原始数据的标准时间并赋给 time_counter
ds_out.coords['time_counter'] = ds_ref['time'].values 

# 重命名经纬度并赋值 (维度变为 y, x)
ds_out['nav_lat'] = (('y', 'x'), ds_ref['lat'].values)
ds_out['nav_lon'] = (('y', 'x'), ds_ref['lon'].values)
ds_out['nav_lat'].attrs['units'] = 'degrees_north'
ds_out['nav_lon'].attrs['units'] = 'degrees_east'

ds_ref.close()

# 循环读取并转换各变量
print(f"开始处理 {target_date} 的数据...")

for target_var, (fname, source_var) in var_mapping.items():
    fpath = os.path.join(base_path, fname)
    ds_in = xr.open_dataset(fpath).sel(time=target_date)
    
    # 提取数组
    data = ds_in[source_var].values
    
    # 执行单位转换
    if target_var == 'tair':
        data = data + 273.15        # deg_C -> K
        units = 'K'
    elif target_var in ['u_wind', 'v_wind']:
        data = data * 0.514444      # kts -> m/s
        units = 'm/s'
    elif target_var == 'atmpres':
        data = data * 100.0         # mb -> Pa
        units = 'Pa'
    elif target_var == 'precip':
        data = data * 1000.0        # m -> kg/m^2 (等效 1mm)
        units = 'kg/m^2'
    elif target_var in ['solar', 'therm_rad']:
        units = 'W/m^2'
    elif target_var == 'qair':
        units = 'kg/kg'
        
    # 存入新 Dataset，维度固定为 (time_counter, y, x)
    ds_out[target_var] = (('time_counter', 'y', 'x'), data)
    ds_out[target_var].attrs['units'] = units
    
    ds_in.close()

# 补充缺失的云覆盖率变量 (使用 time_size 动态适配时间步长度)
ds_out['percentcloud'] = (('time_counter', 'y', 'x'), np.zeros((time_size, y_size, x_size)))
ds_out['percentcloud'].attrs['units'] = 'percent'

# 强制规范时间变量的编码格式
# dtype 设为 float64 或 int64 防止秒数太大溢出（通常 float64 最稳妥）
out_encoding = {
    'time_counter': {
        'units': 'seconds since 1970-01-01 00:00:00',
        'calendar': 'standard',
        'dtype': 'float64'
    }
}

date_str = pd.to_datetime(target_date).strftime('%Ym%md%d')
output_filename = f'CaSR_y{date_str}.nc'

# 导出时传入 encoding 字典
ds_out.to_netcdf(
    output_filename, 
    unlimited_dims=('time_counter',),
    encoding=out_encoding
)

print(f"生成完毕: {output_filename}")

开始处理 2023-03-01 的数据...
生成完毕: CaSR_y2023m03d01.nc


### Inspection

In [2]:
# Inspection

path_CaSR='/home/jqiu/analysis-junqi/Analysis_Atmospheric_Forcing/Data_CaSR_Conversion/CaSR_y2008m01d01.nc'
path_HRDPS='/results/forcing/atmospheric/GEM2.5/operational/ops_y2018m01d01.nc'


In [5]:
# 随便挑一个原文件读取（
ds_test = xr.open_dataset(ref_file)

# 打印 xarray 解析后的属性 和 隐藏的原始单位
print("【当前解析结果】:\n", ds_test['time'].head(5)) 
print("\n【原始文件中的时间单位】:", ds_test['time'].encoding.get('units', '未找到 units 属性'))

【当前解析结果】:
 <xarray.DataArray 'time' (time: 5)> Size: 40B
array(['2007-12-31T13:00:00.000000000', '2007-12-31T14:00:00.000000000',
       '2007-12-31T15:00:00.000000000', '2007-12-31T16:00:00.000000000',
       '2007-12-31T17:00:00.000000000'], dtype='datetime64[ns]')
Coordinates:
  * time     (time) datetime64[ns] 40B 2007-12-31T13:00:00 ... 2007-12-31T17:...
Attributes:
    standard_name:  time
    long_name:      Validity time
    axis:           T

【原始文件中的时间单位】: hours since 2007-12-31T12:00:00


In [7]:
import netCDF4 as nc
import numpy as np

# Define file paths
path_CaSR = '/home/jqiu/analysis-junqi/Analysis_Atmospheric_Forcing/Data_CaSR_Conversion/CaSR_y2008m01d01.nc'
path_HRDPS = '/results/forcing/atmospheric/GEM2.5/operational/ops_y2018m01d01.nc'

# Target variables for comparison
TARGET_VARS = [
    'time_counter', 'x', 'y', 
    'nav_lat', 'nav_lon', 
    'atmpres', 'percentcloud', 'precip', 
    'qair', 'solar', 'tair', 
    'therm_rad', 'u_wind', 'v_wind'
]

def get_var_info(dataset, var_name):
    """Extract attributes and sample data for a single variable, returning a dictionary."""
    if var_name not in dataset.variables:
        return {"shape": "Missing", "type": "Missing", "units": "Missing", "sample": "Missing"}
    
    var_obj = dataset.variables[var_name]
    info = {
        "shape": str(var_obj.shape),
        "type": str(var_obj.datatype),
        "units": getattr(var_obj, 'units', 'None')  # Display "None" if units attribute is missing
    }
    
    # Safely read a small amount of sample data
    if np.prod(var_obj.shape) > 0:
        try:
            if len(var_obj.shape) == 1:
                sample = var_obj[:3]
            elif len(var_obj.shape) == 2:
                sample = var_obj[0, :3]
            elif len(var_obj.shape) == 3:
                sample = var_obj[0, 0, :3]
            elif len(var_obj.shape) == 4:
                sample = var_obj[0, 0, 0, :3]
            else:
                sample = var_obj[...].flatten()[:3]
            
            # Format array: truncate floats to avoid long strings, take first 3 values
            if sample.dtype.kind == 'f':
                sample_str = "[" + ", ".join([f"{val:.6g}" for val in sample]) + ", ...]"
            else:
                sample_str = "[" + ", ".join([str(val) for val in sample]) + ", ...]"
                
            info["sample"] = sample_str
        except Exception:
            info["sample"] = "Read Error"
    else:
        info["sample"] = "Empty Array"
        
    return info

def compare_nc_files_as_table(file1, file2, vars_to_check):
    print("\nOpening files for side-by-side comparison...")
    try:
        ds1 = nc.Dataset(file1, 'r')
        ds2 = nc.Dataset(file2, 'r')
    except Exception as e:
        print(f"Failed to open files: {e}")
        return
    
    # Set table column widths
    col1_w = 14  # Variable Name
    col2_w = 10  # Attribute
    col3_w = 36  # File 1 Data
    col4_w = 36  # File 2 Data
    
    # Print table header
    sep_line = "-" * (col1_w + col2_w + col3_w + col4_w + 13)
    print(sep_line)
    print(f"| {'Variable':<{col1_w}} | {'Attribute':<{col2_w}} | {'CaSR (File 1)':<{col3_w}} | {'HRDPS (File 2)':<{col4_w}} |")
    print(sep_line)
    
    # Iterate through and compare each variable
    for var in vars_to_check:
        info1 = get_var_info(ds1, var)
        info2 = get_var_info(ds2, var)
        
        # First row: Variable Name and Shape
        print(f"| {var:<{col1_w}} | {'Shape':<{col2_w}} | {info1['shape']:<{col3_w}} | {info2['shape']:<{col4_w}} |")
        # Subsequent rows: Type, Units, Sample (Variable column left blank for clean look)
        print(f"| {'':<{col1_w}} | {'Type':<{col2_w}} | {info1['type']:<{col3_w}} | {info2['type']:<{col4_w}} |")
        print(f"| {'':<{col1_w}} | {'Units':<{col2_w}} | {info1['units']:<{col3_w}} | {info2['units']:<{col4_w}} |")
        print(f"| {'':<{col1_w}} | {'Sample':<{col2_w}} | {info1['sample']:<{col3_w}} | {info2['sample']:<{col4_w}} |")
        print(sep_line)

    ds1.close()
    ds2.close()

# Execute comparison
compare_nc_files_as_table(path_CaSR, path_HRDPS, TARGET_VARS)


Opening files for side-by-side comparison...
-------------------------------------------------------------------------------------------------------------
| Variable       | Attribute  | CaSR (File 1)                        | HRDPS (File 2)                       |
-------------------------------------------------------------------------------------------------------------
| time_counter   | Shape      | (24,)                                | (24,)                                |
|                | Type       | float64                              | float64                              |
|                | Units      | seconds since 1970-01-01             | seconds since 1970-01-01 00:00:00.0 0:00 |
|                | Sample     | [1.19915e+09, 1.19915e+09, 1.19915e+09, ...] | [1.51476e+09, 1.51477e+09, 1.51477e+09, ...] |
-------------------------------------------------------------------------------------------------------------
| x              | Shape      | (70,)                 



| Variable | Attribute | CaSR (File 1) | HRDPS (File 2) |
| --- | --- | --- | --- |
| **time_counter** | Shape | (24,) | (24,) |
|  | Type | float64 | float64 |
|  | Units | seconds since 1970-01-01 | seconds since 1970-01-01 00:00:00.0 0:00 |
|  | Sample | [0, 1, 2, ...] | [1.51476e+09, 1.51477e+09, 1.51477e+09, ...] |
| **x** | Shape | (70,) | (256,) |
|  | Type | int64 | float64 |
|  | Units | None | m |
|  | Sample | [0, 1, 2, ...] | [0, 2500, 5000, ...] |
| **y** | Shape | (105,) | (266,) |
|  | Type | int64 | float64 |
|  | Units | None | m |
|  | Sample | [0, 1, 2, ...] | [0, 2500, 5000, ...] |
| **nav_lat** | Shape | (105, 70) | (266, 256) |
|  | Type | float32 | float64 |
|  | Units | degrees_north | degrees_north |
|  | Sample | [43.3772, 43.4174, 43.4575, ...] | [45.6649, 45.6699, 45.6749, ...] |
| **nav_lon** | Shape | (105, 70) | (266, 256) |
|  | Type | float32 | float64 |
|  | Units | degrees_east | degrees_east |
|  | Sample | [235.872, 235.981, 236.09, ...] | [233.016, 233.044, 233.073, ...] |
| **atmpres** | Shape | (24, 105, 70) | (24, 266, 256) |
|  | Type | float32 | float32 |
|  | Units | Pa | Pa |
|  | Sample | [102580, 102618, 102678, ...] | [101921, 101922, 101924, ...] |
| **percentcloud** | Shape | (24, 105, 70) | (24, 266, 256) |
|  | Type | float64 | float32 |
|  | Units | percent | percent |
|  | Sample | [0, 0, 0, ...] | [0, 0, 0, ...] |
| **precip** | Shape | (24, 105, 70) | (24, 266, 256) |
|  | Type | float32 | float32 |
|  | Units | kg/m^2 | kg/m^2 |
|  | Sample | [0, 0, 0, ...] | [0, 0, 0, ...] |
| **qair** | Shape | (24, 105, 70) | (24, 266, 256) |
|  | Type | float32 | float32 |
|  | Units | kg/kg | kg/kg |
|  | Sample | [0.00344162, 0.00322179, 0.00316314, ...] | [0.00547139, 0.00547889, 0.00548639, ...] |
| **solar** | Shape | (24, 105, 70) | (24, 266, 256) |
|  | Type | float32 | float32 |
|  | Units | W/m^2 | W/m^2 |
|  | Sample | [78.8438, 80.5781, 80.4062, ...] | [151.242, 151.242, 151.242, ...] |
| **tair** | Shape | (24, 105, 70) | (24, 266, 256) |
|  | Type | float32 | float32 |
|  | Units | K | K |
|  | Sample | [279.215, 278.611, 277.338, ...] | [283.3, 283.275, 283.25, ...] |
| **therm_rad** | Shape | (24, 105, 70) | (24, 266, 256) |
|  | Type | float32 | float32 |
|  | Units | W/m^2 | W/m^2 |
|  | Sample | [255.942, 254.552, 250.036, ...] | [261.088, 263.311, 263.311, ...] |
| **u_wind** | Shape | (24, 105, 70) | (24, 266, 256) |
|  | Type | float32 | float32 |
|  | Units | m/s | m/s |
|  | Sample | [-2.30252, -2.59893, -2.83304, ...] | [-5.36854, -5.41854, -5.46854, ...] |
| **v_wind** | Shape | (24, 105, 70) | (24, 266, 256) |
|  | Type | float32 | float32 |
|  | Units | m/s | m/s |
|  | Sample | [0.496064, 0.33731, 0.354391, ...] | [-0.907041, -0.877041, -0.837041, ...] |

## Full conversion

Do not run it yet! Not until the `2008-01-01` data are confirmed.

In [ ]:
import xarray as xr
import numpy as np
import pandas as pd
import os
import time
from concurrent.futures import ProcessPoolExecutor, as_completed

def process_single_date(target_date, base_path, output_dir):
    """
    工作进程：处理单一日期的数据。
    (每个进程独立打开文件，避免多进程共享 NetCDF 句柄导致的崩溃)
    """
    date_str_out = target_date.strftime('%Ym%md%d')
    date_query = target_date.strftime('%Y-%m-%d')
    output_filename = os.path.join(output_dir, f'CaSR_y{date_str_out}.nc')

    # 断点续传：如果文件已存在，直接跳过
    if os.path.exists(output_filename):
        return f"跳过 (已存在): {date_query}"

    var_mapping = {
        'solar':     ('Integrated_RDPS_P_FB_SFC_2008_2024.nc',  'CaSR_v3.2_P_FB_SFC'),
        'precip':    ('Integrated_RDPS_P_PR0_SFC_2008_2024.nc', 'CaSR_v3.2_P_PR0_SFC'),
        'therm_rad': ('Integrated_RDPS_P_FI_SFC_2008_2024.nc',  'CaSR_v3.2_P_FI_SFC'),
        'tair':      ('Integrated_RDPS_P_TT_1.5m_2008_2024.nc', 'CaSR_v3.2_P_TT_1.5m'),
        'qair':      ('Integrated_RDPS_P_HU_1.5m_2008_2024.nc', 'CaSR_v3.2_P_HU_1.5m'),
        'u_wind':    ('Integrated_RDPS_P_UUC_10m_2008_2024.nc', 'CaSR_v3.2_P_UUC_10m'),
        'atmpres':   ('Integrated_RDPS_P_PN_SFC_2008_2024.nc',  'CaSR_v3.2_P_PN_SFC'),
        'v_wind':    ('Integrated_RDPS_P_VVC_10m_2008_2024.nc', 'CaSR_v3.2_P_VVC_10m')
    }

    try:
        ds_out = xr.Dataset()

        # 1. 提取网格基准和时间信息 (用当天的数据提取，耗时极短)
        ref_file = os.path.join(base_path, 'Integrated_RDPS_P_TT_1.5m_2008_2024.nc')
        with xr.open_dataset(ref_file) as ds_ref_full:
            ds_ref = ds_ref_full.sel(time=date_query)
            y_size, x_size = ds_ref.rlat.size, ds_ref.rlon.size
            lat_matrix = ds_ref['lat'].values
            lon_matrix = ds_ref['lon'].values
            time_values = ds_ref['time'].values  # 提取真实时间
            time_size = ds_ref.time.size         # 动态获取时间步数

        ds_out.coords['y'] = np.arange(y_size)
        ds_out.coords['x'] = np.arange(x_size)
        ds_out.coords['time_counter'] = time_values # 赋真实 datetime64 值
        
        ds_out['nav_lat'] = (('y', 'x'), lat_matrix)
        ds_out['nav_lon'] = (('y', 'x'), lon_matrix)
        ds_out['nav_lat'].attrs['units'] = 'degrees_north'
        ds_out['nav_lon'].attrs['units'] = 'degrees_east'

        # 2. 循环处理各变量
        for target_var, (fname, source_var) in var_mapping.items():
            fpath = os.path.join(base_path, fname)
            
            # 使用 with 语句确保读取后立刻释放文件锁
            with xr.open_dataset(fpath) as ds_in_full:
                day_slice = ds_in_full[source_var].sel(time=date_query).values

            # 物理单位转换
            if target_var == 'tair':
                data, units = day_slice + 273.15, 'K'
            elif target_var in ['u_wind', 'v_wind']:
                data, units = day_slice * 0.514444, 'm/s'
            elif target_var == 'atmpres':
                data, units = day_slice * 100.0, 'Pa'
            elif target_var == 'precip':
                data, units = day_slice * 1000.0, 'kg/m^2'
            elif target_var in ['solar', 'therm_rad']:
                data, units = day_slice, 'W/m^2'
            elif target_var == 'qair':
                data, units = day_slice, 'kg/kg'
                
            ds_out[target_var] = (('time_counter', 'y', 'x'), data)
            ds_out[target_var].attrs['units'] = units

        # 3. 添加云覆盖 (动态适配时间维大小)
        ds_out['percentcloud'] = (('time_counter', 'y', 'x'), np.zeros((time_size, y_size, x_size), dtype=np.float32))
        ds_out['percentcloud'].attrs['units'] = 'percent'

        # 4. 导出 (添加 unlimited_dims 和 encoding)
        out_encoding = {
            'time_counter': {
                'units': 'seconds since 1970-01-01 00:00:00',
                'calendar': 'standard',
                'dtype': 'float64'
            }
        }
        
        ds_out.to_netcdf(
            output_filename, 
            format='NETCDF4', 
            unlimited_dims=('time_counter',), 
            encoding=out_encoding
        )
        return f"成功生成: {date_query}"

    except Exception as e:
        return f"失败 {date_query} 报错: {str(e)}"

def run_parallel_conversion(base_path, output_dir, start_date='2008-01-01', end_date='2024-12-31', max_workers=8):
    os.makedirs(output_dir, exist_ok=True)
    date_range = pd.date_range(start=start_date, end=end_date, freq='D')
    total_days = len(date_range)
    
    print(f"开始并行处理 {total_days} 天的数据，使用 {max_workers} 个进程...")
    start_time = time.time()

    success_count = 0
    # 使用进程池
    with ProcessPoolExecutor(max_workers=max_workers) as executor:
        # 提交所有任务到进程池
        futures = {
            executor.submit(process_single_date, date, base_path, output_dir): date 
            for date in date_range
        }
        
        # 实时获取完成状态
        for i, future in enumerate(as_completed(futures), 1):
            result = future.result()
            
            if "成功" in result or "跳过" in result:
                success_count += 1
            
            # 每完成 50 个打印一次进度，防止日志刷屏
            if i % 50 == 0 or i == total_days:
                elapsed = time.time() - start_time
                speed = i / elapsed
                print(f"[{i}/{total_days}] {result} (平均速度: {speed:.2f} 天/秒)")

    total_time = time.time() - start_time
    print(f"\n全部完成！耗时: {total_time/60:.2f} 分钟。成功处理: {success_count}/{total_days} 天。")

if __name__ == "__main__":
    SOURCE_PATH = '/ocean/jqiu/forcings_raw/CaSR_raw/2008_2024_Integrated/'
    OUTPUT_PATH = '/ocean/jqiu/forcings/CaSR/'
    
    # max_workers 建议设置为 CPU 核心数的 50%~80%
    run_parallel_conversion(
        base_path=SOURCE_PATH, 
        output_dir=OUTPUT_PATH,
        start_date='2008-01-01', 
        end_date='2024-12-31',
        max_workers=6  # <--- 这里可以根据服务器性能调整
    )

开始并行处理 6210 天的数据，使用 6 个进程...


In [3]:
import os
import numpy as np
import pandas as pd
import xarray as xr


# ============================================================
# 用户配置
# ============================================================

# 输入文件所在目录
BASE_PATH = '/ocean/jqiu/forcings_raw/CaSR_raw/2008_2024_Integrated/'

# 输出目录
OUTPUT_DIR = '/ocean/jqiu/forcings/CaSR/'

# 转换日期范围，包含首尾日期
START_DATE = '2023-03-03'
END_DATE = '2023-03-31'

# 是否覆盖已经存在的输出文件
OVERWRITE_EXISTING = True

# 是否在写出后重新打开文件进行检查
VERIFY_OUTPUT = True


# ============================================================
# 变量映射
# 目标变量名: (输入文件名, 输入变量名)
# ============================================================

VAR_MAPPING = {
    'solar': (
        'Integrated_RDPS_P_FB_SFC_2008_2024.nc',
        'CaSR_v3.2_P_FB_SFC'
    ),
    'precip': (
        'Integrated_RDPS_P_PR0_SFC_2008_2024.nc',
        'CaSR_v3.2_P_PR0_SFC'
    ),
    'therm_rad': (
        'Integrated_RDPS_P_FI_SFC_2008_2024.nc',
        'CaSR_v3.2_P_FI_SFC'
    ),
    'tair': (
        'Integrated_RDPS_P_TT_1.5m_2008_2024.nc',
        'CaSR_v3.2_P_TT_1.5m'
    ),
    'qair': (
        'Integrated_RDPS_P_HU_1.5m_2008_2024.nc',
        'CaSR_v3.2_P_HU_1.5m'
    ),
    'u_wind': (
        'Integrated_RDPS_P_UUC_10m_2008_2024.nc',
        'CaSR_v3.2_P_UUC_10m'
    ),
    'atmpres': (
        'Integrated_RDPS_P_PN_SFC_2008_2024.nc',
        'CaSR_v3.2_P_PN_SFC'
    ),
    'v_wind': (
        'Integrated_RDPS_P_VVC_10m_2008_2024.nc',
        'CaSR_v3.2_P_VVC_10m'
    )
}


# ============================================================
# 辅助函数
# ============================================================

def convert_units(data_array, target_var):
    """
    对指定变量进行单位转换。

    参数
    ----
    data_array : xarray.DataArray
        输入数据。

    target_var : str
        输出变量名称。

    返回
    ----
    converted : xarray.DataArray
        完成单位转换的数据。

    units : str
        转换后的单位。
    """

    if target_var == 'tair':
        # 摄氏度 -> K
        converted = data_array + 273.15
        units = 'K'

    elif target_var in ('u_wind', 'v_wind'):
        # knots -> m/s
        converted = data_array * 0.514444
        units = 'm/s'

    elif target_var == 'atmpres':
        # hPa 或 mb -> Pa
        converted = data_array * 100.0
        units = 'Pa'

    elif target_var == 'precip':
        # m water equivalent -> kg/m²
        converted = data_array * 1000.0
        units = 'kg/m^2'

    elif target_var in ('solar', 'therm_rad'):
        converted = data_array
        units = 'W/m^2'

    elif target_var == 'qair':
        converted = data_array
        units = 'kg/kg'

    else:
        converted = data_array
        units = ''

    return converted, units


def summarize_array(array):
    """
    返回数组的基本统计信息。

    对于全 NaN 数组，min 和 max 返回 NaN。
    """

    values = np.asarray(array)

    total_count = values.size
    finite_mask = np.isfinite(values)
    finite_count = int(finite_mask.sum())
    nan_count = int(np.isnan(values).sum())

    if finite_count > 0:
        minimum = float(np.nanmin(values))
        maximum = float(np.nanmax(values))
        mean = float(np.nanmean(values))
    else:
        minimum = np.nan
        maximum = np.nan
        mean = np.nan

    return {
        'total': total_count,
        'finite': finite_count,
        'nan': nan_count,
        'min': minimum,
        'max': maximum,
        'mean': mean
    }


def print_summary(variable_name, array, indent='    '):
    """
    打印变量统计信息。
    """

    stats = summarize_array(array)

    print(
        f'{indent}{variable_name}: '
        f'shape={np.shape(array)}, '
        f'finite={stats["finite"]}/{stats["total"]}, '
        f'NaN={stats["nan"]}, '
        f'min={stats["min"]:.8g}, '
        f'max={stats["max"]:.8g}, '
        f'mean={stats["mean"]:.8g}'
    )


def select_one_day(dataset, target_date):
    """
    从 Dataset 中选择指定日期。

    使用半开区间 [当天 00:00, 次日 00:00)，
    相比直接 sel(time='YYYY-MM-DD') 更明确。
    """

    day_start = pd.Timestamp(target_date)
    day_end = day_start + pd.Timedelta(days=1)

    return dataset.sel(
        time=slice(
            np.datetime64(day_start),
            np.datetime64(day_end - pd.Timedelta(nanoseconds=1))
        )
    )


def get_output_path(target_date):
    """
    根据日期构造输出文件路径。
    """

    date_str = pd.Timestamp(target_date).strftime('%Ym%md%d')
    filename = f'CaSR_y{date_str}.nc'

    return os.path.join(OUTPUT_DIR, filename)


def verify_written_file(
    output_path,
    expected_time,
    expected_shape
):
    """
    重新打开写出的 NetCDF 文件并进行基本验证。
    """

    print('  正在验证输出文件...')

    with xr.open_dataset(output_path) as ds_check:

        required_dimensions = {
            'time_counter',
            'y',
            'x'
        }

        missing_dimensions = (
            required_dimensions - set(ds_check.dims)
        )

        if missing_dimensions:
            raise ValueError(
                f'输出文件缺少维度: '
                f'{sorted(missing_dimensions)}'
            )

        written_time = ds_check['time_counter'].values

        if not np.array_equal(written_time, expected_time):
            raise ValueError(
                '输出文件中的 time_counter '
                '与参考时间不一致'
            )

        variables_to_check = list(VAR_MAPPING.keys()) + [
            'percentcloud'
        ]

        for variable_name in variables_to_check:

            if variable_name not in ds_check:
                raise KeyError(
                    f'输出文件中缺少变量 '
                    f'{variable_name}'
                )

            variable = ds_check[variable_name]

            if variable.shape != expected_shape:
                raise ValueError(
                    f'{variable_name} 输出形状为 '
                    f'{variable.shape}，'
                    f'预期为 {expected_shape}'
                )

            # 这里触发真正读取文件内容
            values = variable.values

            print_summary(
                variable_name,
                values,
                indent='    验证 '
            )

            if variable_name != 'percentcloud':
                finite_count = int(
                    np.isfinite(values).sum()
                )

                if finite_count == 0:
                    raise ValueError(
                        f'{variable_name} 写出后没有任何'
                        f'有限值，变量可能全部为 NaN'
                    )

    print('  输出文件验证通过。')


# ============================================================
# 单日转换
# ============================================================

def convert_one_day(target_date):
    """
    转换指定日期的数据，每天生成一个 NetCDF 文件。
    """

    print()
    print('=' * 78)
    print(f'开始处理日期: {target_date}')

    output_path = get_output_path(target_date)

    if os.path.exists(output_path):
        if OVERWRITE_EXISTING:
            print(f'输出文件已存在，将覆盖: {output_path}')
        else:
            print(f'输出文件已存在，跳过: {output_path}')
            return 'skipped'

    # --------------------------------------------------------
    # 使用温度文件获取参考时间、网格和经纬度
    # --------------------------------------------------------

    reference_filename = (
        'Integrated_RDPS_P_TT_1.5m_2008_2024.nc'
    )

    reference_path = os.path.join(
        BASE_PATH,
        reference_filename
    )

    if not os.path.exists(reference_path):
        raise FileNotFoundError(
            f'找不到参考文件: {reference_path}'
        )

    with xr.open_dataset(reference_path) as ds_ref_all:

        required_reference_variables = {
            'time',
            'rlat',
            'rlon',
            'lat',
            'lon'
        }

        missing_reference_variables = (
            required_reference_variables
            - set(ds_ref_all.variables)
        )

        if missing_reference_variables:
            raise KeyError(
                f'参考文件缺少变量或坐标: '
                f'{sorted(missing_reference_variables)}'
            )

        ds_ref = select_one_day(
            ds_ref_all,
            target_date
        ).load()

    time_size = ds_ref.sizes.get('time', 0)

    if time_size == 0:
        ds_ref.close()
        print(f'{target_date} 没有参考数据，跳过。')
        return 'no_data'

    y_size = ds_ref.sizes['rlat']
    x_size = ds_ref.sizes['rlon']

    reference_time = ds_ref['time'].values.copy()
    nav_lat = ds_ref['lat'].values.copy()
    nav_lon = ds_ref['lon'].values.copy()

    ds_ref.close()

    expected_shape = (
        time_size,
        y_size,
        x_size
    )

    print(f'时间步数量: {time_size}')
    print(f'网格大小: {y_size} × {x_size}')
    print(f'预期三维数组形状: {expected_shape}')
    print(f'首个时间: {reference_time[0]}')
    print(f'末个时间: {reference_time[-1]}')

    if nav_lat.shape != (y_size, x_size):
        raise ValueError(
            f'lat 形状为 {nav_lat.shape}，'
            f'预期为 {(y_size, x_size)}'
        )

    if nav_lon.shape != (y_size, x_size):
        raise ValueError(
            f'lon 形状为 {nav_lon.shape}，'
            f'预期为 {(y_size, x_size)}'
        )

    # --------------------------------------------------------
    # 创建输出 Dataset
    # --------------------------------------------------------

    ds_out = xr.Dataset(
        coords={
            'time_counter': (
                'time_counter',
                reference_time
            ),
            'y': (
                'y',
                np.arange(y_size, dtype=np.int32)
            ),
            'x': (
                'x',
                np.arange(x_size, dtype=np.int32)
            )
        }
    )

    ds_out['time_counter'].attrs = {
        'standard_name': 'time',
        'long_name': 'Time'
    }

    ds_out['y'].attrs = {
        'long_name': 'Grid y index'
    }

    ds_out['x'].attrs = {
        'long_name': 'Grid x index'
    }

    ds_out['nav_lat'] = (
        ('y', 'x'),
        nav_lat
    )

    ds_out['nav_lon'] = (
        ('y', 'x'),
        nav_lon
    )

    ds_out['nav_lat'].attrs = {
        'standard_name': 'latitude',
        'long_name': 'Latitude',
        'units': 'degrees_north'
    }

    ds_out['nav_lon'].attrs = {
        'standard_name': 'longitude',
        'long_name': 'Longitude',
        'units': 'degrees_east'
    }

    # --------------------------------------------------------
    # 逐个变量读取并转换
    # --------------------------------------------------------

    try:
        for target_var, (
            input_filename,
            source_var
        ) in VAR_MAPPING.items():

            print()
            print(f'  正在处理变量: {target_var}')

            input_path = os.path.join(
                BASE_PATH,
                input_filename
            )

            if not os.path.exists(input_path):
                raise FileNotFoundError(
                    f'找不到输入文件: {input_path}'
                )

            with xr.open_dataset(input_path) as ds_in:

                if source_var not in ds_in:
                    raise KeyError(
                        f'{input_filename} 中不存在变量 '
                        f'{source_var}'
                    )

                source_data = select_one_day(
                    ds_in[[source_var]],
                    target_date
                )[source_var].load()

            source_time_size = source_data.sizes.get(
                'time',
                0
            )

            if source_time_size == 0:
                raise ValueError(
                    f'{target_var} 在 {target_date} '
                    f'没有数据'
                )

            required_dims = {
                'time',
                'rlat',
                'rlon'
            }

            missing_dims = (
                required_dims
                - set(source_data.dims)
            )

            if missing_dims:
                raise ValueError(
                    f'{target_var} 缺少维度: '
                    f'{sorted(missing_dims)}；'
                    f'实际维度为 {source_data.dims}'
                )

            # 先将维度顺序固定下来
            source_data = source_data.transpose(
                'time',
                'rlat',
                'rlon'
            )

            if source_data.shape != expected_shape:
                raise ValueError(
                    f'{target_var} 在 {target_date} '
                    f'的形状为 {source_data.shape}，'
                    f'参考形状为 {expected_shape}'
                )

            source_time = source_data['time'].values

            if not np.array_equal(
                source_time,
                reference_time
            ):
                raise ValueError(
                    f'{target_var} 在 {target_date} '
                    f'的时间坐标与参考文件不一致'
                )

            print_summary(
                f'{target_var} 转换前',
                source_data.values
            )

            converted_data, output_units = convert_units(
                source_data,
                target_var
            )

            # ------------------------------------------------
            # 关键修复：
            #
            # 只取 NumPy 数组，然后按位置写入 ds_out。
            # 不把原始 rlat/rlon 坐标带入输出 Dataset，
            # 从而避免 xarray 按标签自动对齐后产生全 NaN。
            # ------------------------------------------------

            converted_values = np.asarray(
                converted_data.values
            )

            if converted_values.shape != expected_shape:
                raise ValueError(
                    f'{target_var} 转换后的形状为 '
                    f'{converted_values.shape}，'
                    f'预期为 {expected_shape}'
                )

            print_summary(
                f'{target_var} 转换后',
                converted_values
            )

            finite_count = int(
                np.isfinite(converted_values).sum()
            )

            if finite_count == 0:
                raise ValueError(
                    f'{target_var} 转换后没有任何有限值，'
                    f'请检查原始数据'
                )

            ds_out[target_var] = (
                ('time_counter', 'y', 'x'),
                converted_values
            )

            ds_out[target_var].attrs = {
                'units': output_units,
                'coordinates': 'nav_lat nav_lon'
            }

            # 写入 Dataset 后再次确认
            assigned_values = ds_out[target_var].values

            if assigned_values.shape != expected_shape:
                raise ValueError(
                    f'{target_var} 写入 ds_out 后形状为 '
                    f'{assigned_values.shape}，'
                    f'预期为 {expected_shape}'
                )

            assigned_finite_count = int(
                np.isfinite(assigned_values).sum()
            )

            if assigned_finite_count == 0:
                raise ValueError(
                    f'{target_var} 写入 ds_out 后全部为 '
                    f'NaN 或 Inf'
                )

            print_summary(
                f'{target_var} 写入后',
                assigned_values
            )

        # ----------------------------------------------------
        # 添加云覆盖率
        # ----------------------------------------------------

        print()
        print('  正在添加变量: percentcloud')

        percentcloud_values = np.zeros(
            expected_shape,
            dtype=np.float32
        )

        ds_out['percentcloud'] = (
            ('time_counter', 'y', 'x'),
            percentcloud_values
        )

        ds_out['percentcloud'].attrs = {
            'long_name': 'Cloud cover',
            'units': 'percent',
            'coordinates': 'nav_lat nav_lon'
        }

        # ----------------------------------------------------
        # Dataset 全局属性
        # ----------------------------------------------------

        ds_out.attrs = {
            'title': 'Converted CaSR atmospheric forcing',
            'source': (
                'Integrated RDPS/CaSR forcing files'
            ),
            'date_converted': str(target_date),
            'history': (
                'Converted with xarray; spatial dimensions '
                'stored as integer y/x indices.'
            )
        }

        # ----------------------------------------------------
        # 输出编码
        # ----------------------------------------------------

        output_encoding = {
            'time_counter': {
                'units': (
                    'seconds since 1970-01-01 00:00:00'
                ),
                'calendar': 'standard',
                'dtype': 'float64'
            },

            'y': {
                'dtype': 'int32'
            },

            'x': {
                'dtype': 'int32'
            },

            'nav_lat': {
                'zlib': True,
                'complevel': 4
            },

            'nav_lon': {
                'zlib': True,
                'complevel': 4
            }
        }

        for variable_name in VAR_MAPPING:
            output_encoding[variable_name] = {
                'zlib': True,
                'complevel': 4
            }

        output_encoding['percentcloud'] = {
            'zlib': True,
            'complevel': 4,
            'dtype': 'float32'
        }

        # ----------------------------------------------------
        # 写出前进行最终检查
        # ----------------------------------------------------

        print()
        print('  写出前变量统计:')

        for variable_name in (
            list(VAR_MAPPING.keys())
            + ['percentcloud']
        ):
            print_summary(
                variable_name,
                ds_out[variable_name].values
            )

        # ----------------------------------------------------
        # 写出文件
        # ----------------------------------------------------

        print()
        print(f'  正在写出: {output_path}')

        ds_out.to_netcdf(
            output_path,
            mode='w',
            unlimited_dims=['time_counter'],
            encoding=output_encoding
        )

    finally:
        ds_out.close()

    # --------------------------------------------------------
    # 写出后验证
    # --------------------------------------------------------

    if VERIFY_OUTPUT:
        verify_written_file(
            output_path=output_path,
            expected_time=reference_time,
            expected_shape=expected_shape
        )

    print(f'生成完毕: {output_path}')

    return 'success'


# ============================================================
# 日期范围转换
# ============================================================

def convert_date_range(start_date, end_date):
    """
    从 start_date 到 end_date 逐日转换。
    日期范围包含首日和末日。
    """

    os.makedirs(
        OUTPUT_DIR,
        exist_ok=True
    )

    start_time = pd.Timestamp(start_date)
    end_time = pd.Timestamp(end_date)

    if end_time < start_time:
        raise ValueError(
            f'结束日期 {end_date} 不能早于'
            f'开始日期 {start_date}'
        )

    dates = pd.date_range(
        start=start_time,
        end=end_time,
        freq='D'
    )

    print('=' * 78)
    print(f'总共需要处理 {len(dates)} 天')
    print(f'日期范围: {start_date} 至 {end_date}')
    print(f'输入目录: {BASE_PATH}')
    print(f'输出目录: {OUTPUT_DIR}')
    print(f'覆盖已有文件: {OVERWRITE_EXISTING}')
    print(f'写出后验证: {VERIFY_OUTPUT}')

    results = {
        'success': [],
        'skipped': [],
        'no_data': [],
        'failed': []
    }

    for date in dates:

        target_date = date.strftime('%Y-%m-%d')

        try:
            status = convert_one_day(target_date)

            if status not in results:
                status = 'success'

            results[status].append(target_date)

        except Exception as error:
            print()
            print('!' * 78)
            print(f'处理失败: {target_date}')
            print(
                f'错误类型: '
                f'{type(error).__name__}'
            )
            print(f'错误信息: {error}')
            print('!' * 78)

            results['failed'].append(
                {
                    'date': target_date,
                    'error_type': type(error).__name__,
                    'error': str(error)
                }
            )

    # --------------------------------------------------------
    # 最终汇总
    # --------------------------------------------------------

    print()
    print('=' * 78)
    print('全部日期处理完成')
    print(f'成功: {len(results["success"])} 天')
    print(f'跳过: {len(results["skipped"])} 天')
    print(f'无数据: {len(results["no_data"])} 天')
    print(f'失败: {len(results["failed"])} 天')

    if results['success']:
        print()
        print('成功日期:')
        for date in results['success']:
            print(f'  {date}')

    if results['skipped']:
        print()
        print('跳过日期:')
        for date in results['skipped']:
            print(f'  {date}')

    if results['no_data']:
        print()
        print('无数据日期:')
        for date in results['no_data']:
            print(f'  {date}')

    if results['failed']:
        print()
        print('失败日期及原因:')

        for item in results['failed']:
            print(
                f'  {item["date"]}: '
                f'{item["error_type"]}: '
                f'{item["error"]}'
            )

    return results


# ============================================================
# 主程序
# ============================================================

if __name__ == '__main__':
    convert_date_range(
        start_date=START_DATE,
        end_date=END_DATE
    )

总共需要处理 29 天
日期范围: 2023-03-03 至 2023-03-31
输入目录: /ocean/jqiu/forcings_raw/CaSR_raw/2008_2024_Integrated/
输出目录: /ocean/jqiu/forcings/CaSR/
覆盖已有文件: True
写出后验证: True

开始处理日期: 2023-03-03
输出文件已存在，将覆盖: /ocean/jqiu/forcings/CaSR/CaSR_y2023m03d03.nc
时间步数量: 24
网格大小: 105 × 70
预期三维数组形状: (24, 105, 70)
首个时间: 2023-03-03T00:00:00.000000000
末个时间: 2023-03-03T23:00:00.000000000

  正在处理变量: solar
    solar 转换前: shape=(24, 105, 70), finite=176400/176400, NaN=0, min=0, max=672.96875, mean=105.33704
    solar 转换后: shape=(24, 105, 70), finite=176400/176400, NaN=0, min=0, max=672.96875, mean=105.33704
    solar 写入后: shape=(24, 105, 70), finite=176400/176400, NaN=0, min=0, max=672.96875, mean=105.33704

  正在处理变量: precip
    precip 转换前: shape=(24, 105, 70), finite=176400/176400, NaN=0, min=0, max=0.0090040229, mean=0.00024843225
    precip 转换后: shape=(24, 105, 70), finite=176400/176400, NaN=0, min=0, max=9.0040226, mean=0.24843225
    precip 写入后: shape=(24, 105, 70), finite=176400/176400, NaN=0, min=0, max=9.0040

### Inspection



In [2]:
import netCDF4 as nc
import numpy as np

# Define file paths
path_CaSR = '/ocean/jqiu/forcings/CaSR/CaSR_y2023m03d02.nc'
path_HRDPS = '/results/forcing/atmospheric/GEM2.5/operational/ops_y2018m03d02.nc'

# Target variables for comparison
TARGET_VARS = [
    'time_counter', 'x', 'y', 
    'nav_lat', 'nav_lon', 
    'atmpres', 'percentcloud', 'precip', 
    'qair', 'solar', 'tair', 
    'therm_rad', 'u_wind', 'v_wind'
]

def get_var_info(dataset, var_name):
    """Extract attributes and sample data for a single variable, returning a dictionary."""
    if var_name not in dataset.variables:
        return {"shape": "Missing", "type": "Missing", "units": "Missing", "sample": "Missing"}
    
    var_obj = dataset.variables[var_name]
    info = {
        "shape": str(var_obj.shape),
        "type": str(var_obj.datatype),
        "units": getattr(var_obj, 'units', 'None')  # Display "None" if units attribute is missing
    }
    
    # Safely read a small amount of sample data
    if np.prod(var_obj.shape) > 0:
        try:
            if len(var_obj.shape) == 1:
                sample = var_obj[:3]
            elif len(var_obj.shape) == 2:
                sample = var_obj[0, :3]
            elif len(var_obj.shape) == 3:
                sample = var_obj[0, 0, :3]
            elif len(var_obj.shape) == 4:
                sample = var_obj[0, 0, 0, :3]
            else:
                sample = var_obj[...].flatten()[:3]
            
            # Format array: truncate floats to avoid long strings, take first 3 values
            if sample.dtype.kind == 'f':
                sample_str = "[" + ", ".join([f"{val:.6g}" for val in sample]) + ", ...]"
            else:
                sample_str = "[" + ", ".join([str(val) for val in sample]) + ", ...]"
                
            info["sample"] = sample_str
        except Exception:
            info["sample"] = "Read Error"
    else:
        info["sample"] = "Empty Array"
        
    return info

def compare_nc_files_as_table(file1, file2, vars_to_check):
    print("\nOpening files for side-by-side comparison...")
    try:
        ds1 = nc.Dataset(file1, 'r')
        ds2 = nc.Dataset(file2, 'r')
    except Exception as e:
        print(f"Failed to open files: {e}")
        return
    
    # Set table column widths
    col1_w = 14  # Variable Name
    col2_w = 10  # Attribute
    col3_w = 36  # File 1 Data
    col4_w = 36  # File 2 Data
    
    # Print table header
    sep_line = "-" * (col1_w + col2_w + col3_w + col4_w + 13)
    print(sep_line)
    print(f"| {'Variable':<{col1_w}} | {'Attribute':<{col2_w}} | {'CaSR (File 1)':<{col3_w}} | {'HRDPS (File 2)':<{col4_w}} |")
    print(sep_line)
    
    # Iterate through and compare each variable
    for var in vars_to_check:
        info1 = get_var_info(ds1, var)
        info2 = get_var_info(ds2, var)
        
        # First row: Variable Name and Shape
        print(f"| {var:<{col1_w}} | {'Shape':<{col2_w}} | {info1['shape']:<{col3_w}} | {info2['shape']:<{col4_w}} |")
        # Subsequent rows: Type, Units, Sample (Variable column left blank for clean look)
        print(f"| {'':<{col1_w}} | {'Type':<{col2_w}} | {info1['type']:<{col3_w}} | {info2['type']:<{col4_w}} |")
        print(f"| {'':<{col1_w}} | {'Units':<{col2_w}} | {info1['units']:<{col3_w}} | {info2['units']:<{col4_w}} |")
        print(f"| {'':<{col1_w}} | {'Sample':<{col2_w}} | {info1['sample']:<{col3_w}} | {info2['sample']:<{col4_w}} |")
        print(sep_line)

    ds1.close()
    ds2.close()

# Execute comparison
compare_nc_files_as_table(path_CaSR, path_HRDPS, TARGET_VARS)


Opening files for side-by-side comparison...
-------------------------------------------------------------------------------------------------------------
| Variable       | Attribute  | CaSR (File 1)                        | HRDPS (File 2)                       |
-------------------------------------------------------------------------------------------------------------
| time_counter   | Shape      | (24,)                                | (24,)                                |
|                | Type       | float64                              | float64                              |
|                | Units      | seconds since 1970-01-01             | seconds since 1970-01-01 00:00:00.0 0:00 |
|                | Sample     | [1.67772e+09, 1.67772e+09, 1.67772e+09, ...] | [1.51995e+09, 1.51995e+09, 1.51996e+09, ...] |
-------------------------------------------------------------------------------------------------------------
| x              | Shape      | (70,)                 